profiles/killbill/src/main/java/org/killbill/billing/server/notifications/PushNotificationListener.java

```diff
@@ -112,13 +137,104 @@ public Response onCompleted(final Response response) throws Exception {
                         }
                     });
             response = futureStatus.get(timeoutSec, TimeUnit.SECONDS);
+        } catch (final TimeoutException toe) {
+            saveRetryPushNotificationInQueue(tenantId, url, body);
+            return false;
         } catch (final Exception e) {
             log.warn("Failed to push notification url='{}', tenantId='{}'", url, tenantId, e);
             return false;
         }
+
+        if ((response.getStatusCode() == 301) || (response.getStatusCode() == 500)) {
+            saveRetryPushNotificationInQueue(tenantId, url, body);
+            return false;
+        }
+
         return response.getStatusCode() >= 200 && response.getStatusCode() < 300;
     }
 
+    public void resendPushNotification(final PushNotificationKey key, final InternalCallContext callContext) throws JsonProcessingException {
+
+        final NotificationJson notification = new NotificationJson(key.getEventType() != null ? key.getEventType().toString() : null,
+                                                                   key.getAccountId() != null ? key.getAccountId().toString() : null,
+                                                                   key.getObjectType() != null ? key.getObjectType().toString() : null,
+                                                                   key.getObjectId() != null ? key.getObjectId().toString() : null);
+        final String body = mapper.writeValueAsString(notification);
+
+        final BoundRequestBuilder builder = httpClient.preparePost(key.getUrl());
+        builder.setBody(body == null ? "{}" : body);
+        builder.addHeader(HTTP_HEADER_CONTENT_TYPE, CONTENT_TYPE_JSON);
+
+        final Response response;
+        try {
+            final ListenableFuture<Response> futureStatus =
+                    builder.execute(new AsyncCompletionHandler<Response>() {
+                        @Override
+                        public Response onCompleted(final Response response) throws Exception {
+                            return response;
+                        }
+                    });
+            response = futureStatus.get(TIMEOUT_NOTIFICATION, TimeUnit.SECONDS);
+        } catch (final TimeoutException toe) {
+            saveRetryPushNotificationInQueue(key);
+            return;
+        } catch (final Exception e) {
+            log.warn("Failed to push notification url='{}', tenantId='{}'", key.getUrl(), key.getTenantId(), e);
+            return;
+        }
+
+        if ((response.getStatusCode() == 301) || (response.getStatusCode() == 500)) {
+            saveRetryPushNotificationInQueue(key);
+        }
+    }
+
+    private void saveRetryPushNotificationInQueue(final UUID tenantId, final String url, final String body) {
+        try {
+            final NotificationJson notificationJson = mapper.readValue(body, NotificationJson.class);
```

Comment:
We have it accessible from `dispatchCallback` - could we pass it along to save a deserialization call? It would probably also remove the need to catch the `IOException`.